metabase/metabase
# Properly format date-time property for snowplow events

For snowplow events that have a date-time property, it's required to use the RFC3339 format [ref](https://json-schema.org/understanding-json-schema/reference/string.html#dates-and-times).

Currently, we're formatting the date using the ISO8601 format, but we're still receiving snowplow events because there are overlaps between the 2 formats (here is a [cool website](https://ijmacd.github.io/rfc3339-iso8601/) the compare the 2 formats)

For example: 
- `2021-12-29T01:58:25.12+01:00` is valid for both formats
- `2021-12-29T01:58:25.123456+01:00` is valid for ISO8601 but invalid for RFC3339 because it only allows 2 digits for sub-seconds.
- `2021-12-29T01:58:25` is valid for ISO8601 but invalid for RFC3339 because it requires an offset component

In the [instance schema](https://github.com/metabase/metabase/blob/11a018fa5ef86cb6b2a0b221b40735e524a58554/snowplow/iglu-client-embedded/schemas/com.metabase/instance/jsonschema/1-1-0#L37), we use a `created_at` property that was defined to have `date-time` format.

Currently we have 2 ways to get value for this property:
- getting the created_at of the 1st user, which is a  `select now()` in the database (this only applies for instance < 0.41.3 )
  - if the app DB is MySQL, this will return a localdatetime rounded to the second `2021-12-29T01:58:25` => this will be invalidated on snowplow because it doesn't have the offset.
  - if the appDB is Postgres, it will return an offsetdatetime like `2021-12-29T01:58:25.12+01:00` or `2021-12-29T01:58:25.123456+01:00` the difference is in the sub-second, which I _think_ will be system-dependent  
- calling `(t/offset-datetime)` and depending on the system, it could return a time that has 2-digits or 6-digits for sub-seconds.

In other words, the reasons why receive snowplow events now are by accident, and I think we're probably missing many events.

In this PR I added a `format-rfc3339` format to make sure the `date-time` properties are formatted correctly.

Background: I accidentally discovered this issue because I was trying to find out why the snowplow wasn't accepting events from my newly added tracking in #23893. So I decided to use a `real` instance on my local machine and send it to snowcatplow, but it doesn't seem to accept _any_ events from my machine, and the reason was the `created_at` property on my machine has 6-digits for sub-second.

**How to test**
Warning: This testing will send events directly to our production snowcatplow database.
We can't test this with micro because seems like it accepts any date-time string we send :) 

1. hard code `is-prod` in config to true in [config.clj](https://github.com/metabase/metabase/blob/master/src/metabase/config.clj#L76)
2. Starts the instance with MB_SNOWPLOW_AVAILABLE=true
3. Go to Admin > Database > Sample Database > Sync database schema
4. Wait 1-12 hours for snowcatplow to process the events ( sorry :'( )
5. Go to [stats](https://stats.metabase.com/question/5111-task-instance) and filter by `Com Metabase Instance 1 Root -> ID` with the id is your `analytics-uuid` in Setting tables, now you should see events with task-name like : `analyze, classify-fields, sync, sync-fks...`
